packages/fxa-auth-server/lib/payments/customer-locations.ts

```diff
@@ -100,79 +104,234 @@ export class CustomerLocations {
       return { ...bestLocation, country };
     }
     // We have one or more locations for this user; get mode location.
-    locations.sort((a, b) => b.count - a.count);
-    locations = locations.map((location) => ({
+    const copy = [...results];
+    // sort mutates
+    results = copy.sort((a, b) => b.count - a.count);
+    results = results.map((location) => ({
       ...location,
-      country: this.mapLongToShortCountryName(location.country),
-      state: this.mapLongToShortStateName(location.country, location.state),
+      country: this.mapLongToShortCountryName(location.country!),
+      state: this.mapLongToShortStateName(location.country!, location.state!),
     }));
-    const modeLocation = locations[0];
+    const modeLocation = results[0];
     // Prioritize country from Stripe customer
     if (modeLocation.country === country) {
       bestLocation = { ...modeLocation, count: 1 };
-    } else if (locations.some((location) => location.country === country)) {
+    } else if (results.some((location) => location.country === country)) {
       bestLocation = {
-        ...locations.find((location) => location.country === country)!,
+        ...results.find((location) => location.country === country)!,
         count: 1,
       };
     } else {
-      this.log.debug('customerLocations.getBestLocationForPayPalUser', {
+      this.log.info('customerLocations.getBestLocationForPayPalUser', {
         message:
           'No location results countries match the Stripe customer country.',
-        uid,
+        customerId: customer.id,
         country,
       });
       bestLocation = { ...bestLocation, country };
     }
     return bestLocation;
   }
+
+  async isPayPalCustomer(customer: Stripe.Customer): Promise<boolean> {
+    if (customer.metadata.paypalAgreementId) {
+      return true;
+    }
+
+    if (customer?.invoice_settings?.default_payment_method) {
+      return false;
+    }
+
+    this.log.debug('CustomerLocations.noPayPalBaId', {
+      message: "User doesn't have paypal BA ID. Checking DB.",
+      customerId: customer.id,
+    });
+
+    // For PayPal users whose subscriptions are no longer active, they will
+    // not have a paypalBillingAgreementId in Stripe customer metadata.
+    const uidBuffer = uuidTransformer.to(customer.metadata.userid);
+    return !!(await PayPalBillingAgreements.query().findOne({
+      uid: uidBuffer,
+    }));
+  }
+
+  async backfillCustomerLocation({
+    limit,
+    isDryRun,
+    delay,
+  }: {
+    limit: number;
+    isDryRun: boolean;
+    delay: number;
+  }): Promise<void> {
+    let count = 0;
+    const pause = async () =>
+      new Promise((resolve) => {
+        setTimeout(resolve, delay);
+      });
+    const lookupCountries = Object.values(
+      COUNTRIES_LONG_NAME_TO_SHORT_NAME_MAP
+    );
+
+    for await (const customer of this.stripeHelper.stripe.customers.list({
+      expand: ['data.invoice_settings.default_payment_method'],
```

Comment:
It'd be quite a bit more efficient to fetch these in batches of 100 or something ab it higher than the default of 10.